diff --git a/after/plugin/compe_dict.vim b/after/plugin/compe_dict.vim
new file mode 100644
index 0000000..d4a7937
--- /dev/null
+++ b/after/plugin/compe_dict.vim
@@ -0,0 +1,9 @@
+if exists('g:loaded_compe_dict')
+  finish
+endif
+let g:loaded_compe_dict = v:true
+
+if exists('g:loaded_compe')
+  lua require'compe'.register_source('dict', require'compe_dict')
+endif
+
diff --git a/doc/compe.txt b/doc/compe.txt
index 0bf8e0c..075ae05 100644
--- a/doc/compe.txt
+++ b/doc/compe.txt
@@ -342,6 +342,7 @@ tags                Tag completion.
 spell               Spell file completion.
 calc                Lua math expressions.
 omni                Omni completion.
+dict                Dictionary files completion.
 
 Neovim-specific: ~
 nvim_lsp            Neovim's builtin LSP completion.
@@ -391,6 +392,7 @@ Vimscript:
   let g:compe.source.tags = v:true
   let g:compe.source.snippets_nvim = v:true
   let g:compe.source.omni = v:true
+  let g:compe.source.dict = v:true
 <
 Lua:
 >
@@ -419,6 +421,7 @@ Lua:
       tags = true;
       snippets_nvim = true;
       omni = true;
+      dict = true;
     };
   }
 <
diff --git a/lua/compe_dict/init.lua b/lua/compe_dict/init.lua
new file mode 100644
index 0000000..05022eb
--- /dev/null
+++ b/lua/compe_dict/init.lua
@@ -0,0 +1,86 @@
+local compe = require("compe")
+
+local Source = {
+    dicts = {}
+}
+
+function Source.new()
+    return setmetatable({}, { __index = Source })
+end
+
+function Source.get_metadata(_)
+    return {
+        priority = 90;
+        dup = 0;
+        menu = '[Dict]';
+    }
+end
+
+function Source.determine(_, context)
+    return compe.helper.determine(context)
+end
+
+function Source._load_dicts(self)
+    -- get dictionary paths
+    local paths = {}
+    for path in string.gmatch(vim.o.dictionary, "([^,]+)") do
+        paths[path] = true
+    end
+
+    -- load missing dictionaries into sources
+    for path,_ in pairs(paths) do
+        if not self.dicts[path] then
+            self.dicts[path] = {}
+            for line in io.lines(path) do
+                if not string.match(line, "^%s*$") then -- skip empty lines
+                    table.insert(self.dicts[path], line)
+                end
+            end
+        end
+    end
+
+    -- removes unused dictionaries
+    for path,_ in pairs(self.dicts) do
+        if not paths[path] then
+            self.dicts[path] = nil
+        end
+    end
+end
+
+function Source._matches(self, word)
+    local matches = {}
+
+    -- set a smartcase pattern for word
+    local pattern = "^" .. string.gsub(word, "%a", function (c)
+        if string.match(c, "%u") then
+            return c
+        else
+            return string.format("[%s%s]", 
+            string.lower(c),
+            string.upper(c))
+        end
+    end)
+
+    for path,dict in pairs(self.dicts) do
+        local filename = vim.fn.fnamemodify(path, ":t:r")
+        for _,line in ipairs(dict) do
+            if string.match(line, pattern) then
+                table.insert(matches, {
+                    word = line,
+                    kind = filename
+                })
+            end
+        end
+    end
+    return matches
+end
+
+function Source.complete(self, context)
+    self:_load_dicts()
+    context.callback({
+        items = self:_matches(context.input)
+    })
+end
+
+return Source.new()
+
